### DISCUSSION TITLE
Help structuring user plugins configuration modules correctly

### DISCUSSION THREAD

**Discussion by lyndhurst** [LunarVim/LunarVim#4125] [Q&A] [answered]
Hi,

I try to split my `lvim` config into separate modules including the user plugins config files.
When I mess up somewhere in my config, `lvim` starts up looking like vanilla `neovim`, that is ok.

The problem is that after fixing the issue, `lvim` won't load configuration anymore. I get an error thrown because my plugin configuration modules call for some `foo.setup()` functions from a `foo` plugin that is not installed / loaded yet.

The only way I found to fix it is to go over the `lvim.plugins` table, comment out all the "config calls", reload `lvim`, uncomment the "config calls" and restart `lvim`. Not very convenient when having several of them !

I am sure there is a more efficient / elegant way to structure my config, but looking around github public config files, I did not find it yet. I describe below my current setup more explicitly.  Some help would be greatly appreciated so that I can build my config from saner foundations. Thanks.

**My current setup**

I add plugins to the `lvim.plugins` table from the file `lua/user/plugins.lua`.
There, I call other modules to configure my plugins. It looks like that:

```lua
local M = {}

M.setup = function()
  lvim.plugins = {
    {
      "folke/trouble.nvim",
      cmd = "TroubleToggle",
    },
    {
      'jedrzejboczar/possession.nvim',
      config = require("user.possession").config()
    },
  }
end

return M
```
The plugin configuration modules look like that

```lua
local M = {}

M.config = function()
  require('possession').setup({
  -- custom configuration
  })
end

return M
```

**Comment by LostInTheLogs** [comment, marked as answer]
```diff
local M = {}

M.setup = function()
  lvim.plugins = {
    {
      "folke/trouble.nvim",
      cmd = "TroubleToggle",
    },
    {
      'jedrzejboczar/possession.nvim',
--    config = require("user.possession").config()
++    config = require("user.possession").config
    },
  }
end

return M
```

**Reply by lyndhurst** [reply]
Thanks for the help, I cannot test right now, but this makes a lot of sense. I will mark the answer as soon as I can.

**Comment by lyndhurst** [comment]
Thanks for the help, I was confused by not being familiar with `lua` and kind of overlooked that I was calling functions instead of passing them as objects.
I can mess up my confguration at will now ! :+1:
